cpp/ql/src/semmle/code/cpp/models/implementations/StdPair.qll

```diff
@@ -0,0 +1,60 @@
+/**
+ * Provides models for the C++ `std::pair` class.
+ */
+
+import semmle.code.cpp.models.interfaces.Taint
+
+/**
+ * Additional model for `std::pair` constructors.
+ */
+class StdPairConstructor extends Constructor, TaintFunction {
+  StdPairConstructor() { this.hasQualifiedName("std", "pair", "pair") }
+
+  /**
+   * Gets the index of a parameter to this function that is a reference to
+   * either value type of the pair.
+   */
+  int getAValueTypeParameterIndex() {
+    getParameter(result).getUnspecifiedType().(ReferenceType).getBaseType() =
+      getDeclaringType().getTemplateArgument(_).(Type).getUnspecifiedType() // i.e. the `T1` or `T2` of this `std::pair<T1, T2>`
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // taint flow from any parameter of a value type to the qualifier
+    input.isParameterDeref(getAValueTypeParameterIndex()) and
+    (
+      output.isReturnValue() // TODO: this is only needed for AST data flow, which treats constructors as returning the new object
+      or
+      output.isQualifierObject()
+    )
+  }
+}
+
+/**
+ * An instantiation of `std::make_pair`.
+ */
+class StdMakePair extends TaintFunction {
+  StdMakePair() { this.hasQualifiedName("std", "make_pair") }
```

Comment:
When we discussed key/value conflation in `std::map` I had not foreseen that we'd also get field conflation in `std::pair`. The latter seems much worse, and I'm tempted to restrict the `std::pair` models introduced here to only be for the template instantiations of `std::pair` that are used with `std::{map,unordered_map}` somewhere in the snapshot. That would be non-local and somewhat chaotic, but it would probably remove a lot of field conflation in practice. What do you think?

Reply:
> I'm tempted to restrict the `std::pair` models introduced here to only be for the template instantiations of `std::pair` that are used with `std::{map,unordered_map}` somewhere in the snapshot.

Whilst this _might_ be an acceptable temporary measure, I don't like it.  There will be pairs of types that are used both for maps and other cases, the result as you point out would be chaotic and non-local (and the QL will be ugly).

An alternative workaround would be to limit taint to the second components of `pair` and `map`.  For `map` this is naturally the part that's most interesting, for `pair` this is arbitrary except for its correspondence with `map`.

I'd prefer not to have to spend time on a workaround at all, and create a proper solution for this problem.  We need to spend more time discussing that though.

Reply:
> An alternative workaround would be to limit taint to the second components of `pair` and `map`. For `map` this is naturally the part that's most interesting, for `pair` this is arbitrary except for its correspondence with `map`.

I think this is the solution I like best, but I expect quite a few users will be confused by this.

Reply:
Done: I've changed the `std::pair` model to only accept taint from parameters that correspond with the `second` value of the pair, so that there should be no conflation between the fields in `pair` (I think).  This change should be undone when we have `TaintFunction` support for tainting the specific corresponding member variables (ticket: **TODO**).

I don't think any changes are required for `std::map` because in all modelled cases, the input is a `std::pair` (so the above change is sufficient), or comes from another `std::map` (same), or in the case of `std::insert_or_assign` already neglects flow from the key.

I also experimented with removing the `make_pair` model in favour of relying on taint flow through the library code (it's a template so the body should always be visible).  This works for a trivial implementation of `make_pair`, but the GCC one I found in the wild is more complicated (using `std::forward` in particular), and I'm not confident that flow will work or that I can fairly test that it will work.  So I've left the model of `make_pair` in for now (but only passing in taint from the second argument).

Reply:
I thought we already modelled `std::forward` as an identity function?

Reply:
So we do.  I suppose that means it wouldn't matter how badly I mess up defining `std::forward` accurately in the test code...

Reply:
> (ticket: **TODO**).

I've opened a https://github.com/github/codeql-c-analysis-team/issues/153 for this.

Reply:
I've now removed the model for `make_pair` as the test shows that it isn't necessary when we have a definition.  Thanks @MathiasVP !